File: Remove_K_digits.cpp
class Solution {
  public:
    string removeKdigits(string S, int K){
        int n = S.length();
        if(n == K){
            return "0";
        }
        stack<char> store;
        for(int i = 0; i < n; i++){
            while(K and !store.empty() and store.top() > S[i]){
                store.pop();
                K--;
            }
            store.push(S[i]);
        }
        while(K){
            store.pop();
            K--;
        }
        string ans = "";
        while(!store.empty()){
            ans.push_back(store.top());
            store.pop();
        }
        reverse(ans.begin(), ans.end());
        for(int i = 0; i < ans.size(); i++){
            if(ans[i] != '0'){
                return ans.substr(i);
            }
        }
        return "0";
    }
};

File: Remove_K_digits.py
class Solution:
  def removeKdigits(self, S, K):
    n = len(S)
    if n == K:
        return "0"
    store = []
    for i in range(n):
        while K and store and store[-1] > S[i]:
            store.pop()
            K -= 1
        store.append(S[i])
    while K:
        store.pop()
        K -= 1
    ans = ""
    while store:
        ans += store.pop()
    ans = ans[::-1]
    for i in range(len(ans)):
        if ans[i] != '0':
            return ans[i:]
    return "0"
